### PR TITLE
MockConnectionHandler & Reader/Writer tests

### PR DESCRIPTION
**Pull request by jamiebrynes7**
#### Description
Added a:

- `MockConnectionHandler`
- `DynamicConverter` for converting snapshot -> update
- Reader/Writer injection and enabling criteria.

#### Tests
Ran tests.

#### Documentation
How is this documented (for example: release note, upgrade guide, feature page, in-code documentation)?
N/A

**Did you remember a changelog entry?**
Yes 

#### Primary reviewers
If your change will take a long time to review, you can name at most two primary reviewers who are ultimately responsible for reviewing this request. @ mention them.

### CODE REVIEW

File: workers/unity/Packages/com.improbable.gdk.core/Worker/ConnectionHandlers/MockConnectionHandler.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            diff.AddEvent(ev, entityId, componentId, updateId++);
+        }
+
+        // TODO: Commands
+
+        #region IConnectionHandler implementation
+
+        public ViewDiff GetMessagesReceived()
+        {
+            return diff;
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.core/Worker/ConnectionHandlers/MockConnectionHandler.cs]: This looks like you edit the diff as you are reading it

Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.core/Worker/ConnectionHandlers/MockConnectionHandler.cs]: Its up to the caller to ensure things happen in the correct order.

Alternatively we can have two diffs that we swap around

File: workers/unity/Packages/com.improbable.gdk.core/Dynamic/DynamicConverter.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
+using System;
+
+namespace Improbable.Gdk.Core
+{
+    public static class DynamicConverter
+    {
+        public delegate TUpdate SnapshotToUpdate<TSnapshot, out TUpdate>(TSnapshot snapshot)
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.core/Dynamic/DynamicConverter.cs]: Can take things by ref

Comment (samiwh) [reply workers/unity/Packages/com.improbable.gdk.core/Dynamic/DynamicConverter.cs]: Or in actually

Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.core/Dynamic/DynamicConverter.cs]: Yes that is true

File: workers/unity/Packages/com.improbable.gdk.core/Worker/ConnectionHandlers/MockConnectionHandler.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+        public void CreateEntity(long entityId, EntityTemplate template)
+        {
+            var handler = new EntityTemplateDynamicHandler(template, entityId, diff);
+            DynamicConverter.ForEachComponent(handler);
+        }
+
+        public void ChangeAuthority(long entityId, uint componentId, Authority newAuthority)
+        {
+            diff.SetAuthority(entityId, componentId, newAuthority);
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.core/Worker/ConnectionHandlers/MockConnectionHandler.cs]: Worth noting - this still requires the user to update the authority in the right order

File: workers/unity/Packages/com.improbable.gdk.core/Worker/ConnectionHandlers/MockConnectionHandler.cs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        }
+
+        public void UpdateComponent<T>(long entityId, uint componentId, T update) where T : ISpatialComponentUpdate
+        {
+            diff.AddComponentUpdate(update, entityId, componentId, updateId++);
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.core/Worker/ConnectionHandlers/MockConnectionHandler.cs]: This means we can't have updates and events arriving atomically

Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.core/Worker/ConnectionHandlers/MockConnectionHandler.cs]: Can add a method to do that

File: workers/unity/Packages/com.improbable.gdk.core/Worker/ConnectionHandlers/MockConnectionHandler.cs (hunk 3 above)
Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.core/Worker/ConnectionHandlers/MockConnectionHandler.cs]: Yes that is true - we could add more safety checks to this in the future if we find its weird

File: test-project/Assets/EditmodeTests/Subscriptions/ReaderWriter/InjectionCriteriaTests.cs [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+            linker.LinkGameObjectToSpatialOSEntity(new EntityId(entityId), gameObject);
+            requireLifecycleSystem.Update();
+
+            return gameObject;
+        }
+
+        private class PositionReaderBehaviour : MonoBehaviour
+        {
+            [Require] public PositionReader Reader;
```
Comment (samiwh) [inline test-project/Assets/EditmodeTests/Subscriptions/ReaderWriter/InjectionCriteriaTests.cs]: #pragma warning disable 649

**Pull request merged**